contrib/mesos/pkg/minion/task.go

```diff
@@ -0,0 +1,575 @@
+
+type taskCompletion struct {
+	name string // name of the task
+	code int    // exit code that the task process completed with
+}
+
+// liveProcess is a useful abstraction for testing
+type liveProcess interface {
+	// Wait works like exec.Cmd.Wait()
+	Wait() error
+
+	// Kill returns the pid of the process that was killed
+	Kill() (int, error)
+}
+
+type cmdProcess struct {
+	delegate *exec.Cmd
+}
+
+func (cp *cmdProcess) Wait() error {
+	return cp.delegate.Wait()
+}
+
+func (cp *cmdProcess) Kill() (int, error) {
+	// TODO(jdef) SIGKILL is pretty harsh, should we implement kill escalation: start with TERM
+	// and follow w/ KILL in x seconds?
+
+	// kill the entire process group, not just the one process
+	pid := cp.delegate.Process.Pid
+	processGroup := 0 - pid
+	rc := syscall.Kill(processGroup, syscall.SIGKILL)
+	return pid, rc
+}
+
+type task struct {
+	name         string                // required: unique name for this task
+	bin          string                // required: path to executable
+	args         []string              // optional: process arguments
+	env          []string              // optional: process environment override
+	createLogger func() io.WriteCloser // factory func that builds a log writer
+	cmd          liveProcess           // process that we started
+	errCh        chan error            // errCh reports errors encountered while trying to execute task processes
+	completedCh  chan *taskCompletion  // code reports exit codes encountered when task processes exit
+	shouldQuit   chan struct{}         // shouldQuit is closed to indicate that the task should stop its running process, if any
+	done         chan struct{}         // done closes when all processes related to the task have terminated
+	finished     func(ok bool) bool    // callback invoked when a task process has completed; if ok then it will be restarted if it returns true
+	restartDelay time.Duration         // interval between repeated task restarts
+	startProcess taskStateFn           // prepare and start a new live process, defaults to notStartedTask; should be set by run()
+}
+
+// makeTask builds a newly initialized task object but does not start any processes for it
+func makeTask(name string, bin string, args []string, env []string, cl func() io.WriteCloser) *task {
+	return &task{
+		name:         name,
+		bin:          bin,
+		args:         args,
+		env:          env,
+		createLogger: cl,
+		completedCh:  make(chan *taskCompletion),
+		errCh:        make(chan error),
+		shouldQuit:   make(chan struct{}),
+		done:         make(chan struct{}),
+		restartDelay: defaultTaskRestartDelay,
+	}
+}
+
+// spawnTask creates a new task and immediately attempts to spawn a process for the task
+func spawnTask(name string, bin string, args []string, env []string, cl func() io.WriteCloser) *task {
+	t := makeTask(name, bin, args, env, cl)
+	go t.run(notStartedTask)
+	return t
+}
+
+func (t *task) run(initialState taskStateFn) {
```

Comment:
docs needed: run executes the state machine responsible for starting, monitoring, and possibly restarting a system process for the task. The initialState func is the entry point of the state machine. Upon returning the done, completedCh, and errCh chans are all closed.